language/stdlib/modules/Roles.move

```diff
@@ -0,0 +1,265 @@
+    /// module. This should hopefully make proofs easier.
+    public fun new_role(
+        creating_account: &signer,
+        new_account: &signer,
+        role_id: u64,
+    ) acquires RoleId {
+        let calling_role = borrow_global<RoleId>(Signer::address_of(creating_account));
+        // A role cannot have previously been assigned to `new_account`.
+        assert(!exists<RoleId>(Signer::address_of(new_account)), 1);
+        if (role_id == VALIDATOR_ROLE_ID() ||
+            role_id == VALIDATOR_OPERATOR_ROLE_ID() ||
+            //role_id == DESIGNATED_DEALER_ROLE_ID() || // Shouldn't this be TC?
+            role_id == PARENT_VASP_ROLE_ID()
+        ) {  // Granted by the association root
+            // calling_role must be the association root role
+            assert(
+                calling_role.role_id == ASSOCIATION_ROOT_ROLE_ID()
+                // XXX/HACK/REMOVE(tzakian): This is for testnet semantics
+                // only. THIS NEEDS TO BE REMOVED.
+                || calling_role.role_id == TREASURY_COMPLIANCE_ROLE_ID(),
+                0
+            );
+            publish_role_privilege_for_role_id(new_account, role_id);
+            move_to(new_account, RoleId { role_id })
+        } else if (role_id == CHILD_VASP_ROLE_ID()) { // granted by parent vasp
+            // calling_role must be a parent vasp role
+            assert(calling_role.role_id == PARENT_VASP_ROLE_ID(), 0);
+            publish_role_privilege_for_role_id(new_account, role_id);
+            move_to(new_account, RoleId { role_id })
+        } else if (role_id == DESIGNATED_DEALER_ROLE_ID()) { // granted by TC
+            // calling_role must be treasury/compliance
+            assert(calling_role.role_id == TREASURY_COMPLIANCE_ROLE_ID(), 0);
+            publish_role_privilege_for_role_id(new_account, role_id);
+            move_to(new_account, RoleId { role_id })
+        } else if (role_id == UNHOSTED_ROLE_ID()) {
+            // TODO(tzakian): remove unhosted creation/guard so that only
+            // assoc root can create.
+            publish_role_privilege_for_role_id(new_account, role_id);
+            move_to(new_account, RoleId { role_id })
+        } else { // Unknown account role
+            // UNRECOGNIZED_ACCOUNT_ROLE
+            abort 2
+        }
+    }
+
+    /// A role comes with a default privilege for its role. This can allow
+    /// extensibility later on if a new module introduces a privilege for a role `R`.
+    /// The new module can use a capability for the role `R`;
+    /// `&Capability<R>` to guard the  granting of the privilege in order
+    /// to ensure that only those with appropriate permissions are granted
+    /// the new permission. e.g.
+    /// ```
+    /// public fun publish_new_privilege(account: &signer, _: &Capability<R>) {
+    ///    Roles::add_privilege_to_account(account, Roles::R_ROLE_ID());
+    /// }
+    ///```
+    fun publish_role_privilege_for_role_id(account: &signer, role_id: u64) {
```

Comment:
Wondering if this would be simpler for the prover if you combined this with `new_role`--that way, it will only  have to reason about the  branching on `role_id` once.